ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java

```diff
@@ -428,6 +430,11 @@ protected Boolean findNextCompactionAndExecute(boolean collectGenericStats, bool
 
         AcidMetricService.updateMetricsFromWorker(ci.dbname, ci.tableName, ci.partName, ci.type,
             dir.getCurrentDirectories().size(), dir.getDeleteDeltas().size(), conf, msc);
+
+        // Get columns with stats at the end.
+        // This will save computations if compaction has failed.
+        columnList = msc.findColumnsWithStats(CompactionInfo.compactionInfoToStruct(ci));
```

Comment:
Why don't we fetch the column list right before we call `statsUpdater.gatherStats`? Also it unnecessary to call it always as it's not cheap, only when auto-stats is disabled and it's a MAJOR compaction. Maybe would be even better to do it in StatsUpdater

Reply:
> Maybe would be even better to do it in StatsUpdater

Implemented this way. Done